server/src/unifyfs_server_pid.c

```diff
@@ -0,0 +1,213 @@
+    hg_handle_t handle;
+    server_pid_in_t in;
+    server_pid_out_t out;
+
+    in.rank = glb_pmi_rank;
+    in.pid = server_pid;
+
+    hret = margo_create(unifyfsd_rpc_context->svr_mid,
+                        glb_servers[0].margo_svr_addr,
+                        unifyfsd_rpc_context->rpcs.server_pid_id,
+                        &handle);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to create rpc handle (ret=%d)", hret);
+        return EIO;
+    }
+
+    hret = margo_forward(handle, &in);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to forward rpc (ret=%d)", hret);
+        return EIO;
+    }
+
+    hret = margo_get_output(handle, &out);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to get rpc result (ret=%d)", hret);
+        return EIO;
+    }
+
+    ret = out.ret;
+
+    return ret;
+}
+
+static void server_pid_handle_rpc(hg_handle_t handle)
+{
+    int ret = 0;
+    int i = 0;
+    int count = 0;
+    hg_return_t hret = 0;
+    server_pid_in_t in;
+    server_pid_out_t out;
+
+    hret = margo_get_input(handle, &in);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to get input (ret=%d)", hret);
+        return;
+    }
+
+    server_pids[in.rank] = in.pid;
+
+    out.ret = 0;
+    hret = margo_respond(handle, &out);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to respond rpc (ret=%d)", hret);
+        return;
+    }
+
+    margo_free_input(handle, &in);
+    margo_destroy(handle);
+
+    for (i = 0; i < glb_pmi_size; i++) {
+        if (server_pids[i] > 0) {
+            count++;
+        }
+    }
+
+    if (count == glb_pmi_size) {
+        ret = pthread_cond_signal(&server_pid_cond);
+        if (ret) {
+            LOGERR("failed to signal condition (%s)", strerror(ret));
+        }
+    }
+}
+DEFINE_MARGO_RPC_HANDLER(server_pid_handle_rpc);
+
+static inline int set_timeout(void)
+{
+    int ret = 0;
+    long timeout_sec = 0;
+
+    if (server_cfg.server_init_timeout) {
+        ret = configurator_int_val(server_cfg.server_init_timeout,
+                                   &timeout_sec);
+        if (ret) {
+            LOGERR("failed to read configuration");
+            return ret;
+        }
+    }
+
+    clock_gettime(CLOCK_REALTIME, &server_pid_timeout);
+    server_pid_timeout.tv_sec += timeout_sec;
+
+    return 0;
+}
+
+static int create_server_pid_file(void)
+{
+    int i = 0;
+    int ret = 0;
+    char filename[PATH_MAX] = { 0, };
+    FILE* fp = NULL;
+
+    if (!server_pids) {
+        LOGERR("cannot access the server pids");
+        ret = EINVAL;
+        goto out;
+    }
+
+    sprintf(filename, "%s/%s", server_cfg.sharedfs_dir, UNIFYFSD_PID_FILENAME);
+
+    fp = fopen(filename, "w");
+    if (!fp) {
+        LOGERR("failed to create file %s (%s)", filename, strerror(errno));
+        ret = errno;
+        goto out;
+    }
+
+    for (i = 0; i < glb_pmi_size; i++) {
+        fprintf(fp, "[%d] %d\n", i, server_pids[i]);
+    }
+
+    fclose(fp);
+
+out:
```

Comment:
you should only use labels when there is common code that must be run at the end of the function in many cases. this function has no such code. replace all your gotos with explicit returns.